x/incentive/types/params_test.go

```diff
@@ -21,129 +22,130 @@ type ParamTestSuite struct {
-				Denom:         "bnb",
-				Reward:        sdk.NewCoin("ukava", sdk.NewInt(10000000000)),
-				Duration:      time.Hour * 24 * 7,
-				TimeLock:      time.Hour * 8766,
-				ClaimDuration: time.Hour * 0,
-			},
-		},
-	}
-	p6 := types.Params{
-		Active: true,
-		Rewards: types.Rewards{
-			types.Reward{
-				Active:        true,
-				Denom:         "bnb",
-				Reward:        sdk.NewCoin("ukava", sdk.NewInt(0)),
-				Duration:      time.Hour * 24 * 7,
-				TimeLock:      time.Hour * 8766,
-				ClaimDuration: time.Hour * 0,
-			},
-		},
-	}
-
 	suite.tests = []paramTest{
 		paramTest{
-			params:     p1,
+			name: "valid",
+			params: types.Params{
+				Active: true,
+				Rewards: types.Rewards{
+					types.Reward{
+						Active:        true,
+						Denom:         "bnb",
+						Reward:        sdk.NewCoin("ukava", sdk.NewInt(10000000000)),
+						Duration:      time.Hour * 24 * 7,
+						TimeLock:      time.Hour * 8766,
+						ClaimDuration: time.Hour * 24 * 14,
+					},
+				},
+			},
 			expectPass: true,
 		},
 		paramTest{
-			params:     p2,
+			name: "duplicate reward",
+			params: types.Params{
+				Active: true,
+				Rewards: types.Rewards{
+					types.Reward{
+						Active:        true,
+						Denom:         "bnb",
+						Reward:        sdk.NewCoin("ukava", sdk.NewInt(10000000000)),
+						Duration:      time.Hour * 24 * 7,
+						TimeLock:      time.Hour * 8766,
+						ClaimDuration: time.Hour * 24 * 14,
+					},
+					types.Reward{
+						Active:        true,
+						Denom:         "bnb",
+						Reward:        sdk.NewCoin("ukava", sdk.NewInt(10000000000)),
+						Duration:      time.Hour * 24 * 7,
+						TimeLock:      time.Hour * 8766,
+						ClaimDuration: time.Hour * 24 * 14,
+					},
+				},
+			},
 			expectPass: false,
 		},
 		paramTest{
-			params:     p3,
+			name: "negative reward duration",
+			params: types.Params{
+				Active: true,
+				Rewards: types.Rewards{
+					types.Reward{
+						Active:        true,
+						Denom:         "bnb",
+						Reward:        sdk.NewCoin("ukava", sdk.NewInt(10000000000)),
+						Duration:      time.Hour * -24 * 7,
+						TimeLock:      time.Hour * 8766,
+						ClaimDuration: time.Hour * 24 * 14,
+					},
+				},
+			},
 			expectPass: false,
 		},
 		paramTest{
-			params:     p4,
+			name: "negative time lock",
+			params: types.Params{
+				Active: true,
+				Rewards: types.Rewards{
+					types.Reward{
+						Active:        true,
+						Denom:         "bnb",
+						Reward:        sdk.NewCoin("ukava", sdk.NewInt(10000000000)),
+						Duration:      time.Hour * 24 * 7,
+						TimeLock:      time.Hour * -8766,
+						ClaimDuration: time.Hour * 24 * 14,
+					},
+				},
+			},
 			expectPass: false,
 		},
 		paramTest{
-			params:     p5,
+			name: "zero claim duration",
+			params: types.Params{
+				Active: true,
+				Rewards: types.Rewards{
+					types.Reward{
+						Active:        true,
+						Denom:         "bnb",
+						Reward:        sdk.NewCoin("ukava", sdk.NewInt(10000000000)),
```

Comment:
Since sdk.NewCoin() panics on an empty denom, could a Reward with `sdk.NewCoin("", sdk.NewInt(10000000000))` be submitted in params via governance? I believe proposals are json - so if the coin is first instantiated within param validation could it cause a panic if `len(denom) > 0` isn't checked?

Reply:
Added a check for `denom == ""`

Reply:
This has me concerned about other coin denoms in other modules' params. We should check that we're validating them appropriately everywhere.

Reply:
It seems like this is an issue that someone else has thought about. I guess it makes sense for all param validation to be handled at the module level, but I'm interested if there any proposal safeguards in place in the Cosmos SDK for preventing `sdk.Coins` from causing a panic @fedekunze